programmazione-concorrente.md

```diff
@@ -1 +1,512 @@
+        // Fai cose
+    }
+}
+...
+ // Fa partire il thread e ritorna subito
+new MyThread().start();
+```
+
+Per motivi di riutilizzo del codice, viene solitamente preferito utilizzare un metodo alternativo per la creazione di
+un Thread, implementando invece l'interfaccia `java.lang.Runnable`:
+```java
+class MyRunnable implements Runnable {
+    @Override
+    public void run() {
+        // Fai cose
+    }
+}
+...
+// Fa partire il thread e ritorna subito
+new MyThread(new MyRunnable()).start();
+```
+
+
+Notiamo come, in entrambi i casi, il thread sia libero di accedere a qualsiasi variabile globale o passata nel costruttore,
+causando potenzialmente errori di interferenza se più thread accedono alle stesse variabili contemporaneamente.
+
+## Intrinsic lock
+
+Java, di default, associa a ogni oggetto un proprio "lock", ovvero un oggetto che permette di implementare mutua esclusione tra
+i vari thread su sezioni di codice di interesse. Questo avviene mediante l'uso della keyword `synchronized`, specificando poi
+l'oggetto di cui si vuole utilizzare il lock:
+```java
+class Account {
+    private float balance;
+
+    public Account(float balanceIniz) { 
+        balance = balanceIniz; 
+    }
+    
+    public void deposit(float money) { 
+        synchronized(this) { 
+            balance += money; 
+        }
+    }
+    public void withdraw(float money) {
+        synchronized(this) {
+            balance -= money; 
+        }
+    }
+}
+```
+
+L'intrinsic lock:
+- mette in mutua esclusione sezioni di codice, permettendo a un solo thread alla volta
+  l'esecuzione, sospendendo gli altri thread chiamanti fino al completamento.
+- é un lock rientrante: se il thread, quando raggiunge una sezione critica, è già in
+  possesso del lock, continuerà la sua esecuzione senza venire sospeso.
+- quando viene acquisito/rilasciato, forza la sincronizzazione della cache locale del thread con 
+  la memoria centrale, garantendo che tutte le variabili lette siano aggiornate.
+
+Notiamo come, a causa della cache locale del thread menzionata al terzo punto, sia necessario
+mettere in sezioni sincronizzate anche metodi che leggono solo una variabile, senza modificarla.
+Questo perché, sebbene il metodo stesso non la modifichi, la variabile potrebbe però essere
+cambiata esternamente da altri thread, causando un desync tra la cache locale e la memoria centrale.
+
+Esiste anche una forma più breve di scrivere il codice precedente, mettendo la keyword
+direttamente sul metodo:
+```java
+class Account {
+    private float balance;
+
+    public Account(float balanceIniz) { 
+        balance = balanceIniz; 
+    }
+    
+    public synchronized void deposit(
+        float money) { 
+        balance += money; 
+    }
+
+    public synchronized void withdraw(
+        float money) {
+        balance -= money; 
+    }
+}
+```
+
+---
+
+In generale, bisogna prestare particolare attenzione nell'uso di questa forma (soprattutto nei TdE), in quanto,
```

Comment:
Scusami, credo di essermi spiegato male. Intendevo fare un esempio dove invece che lockare su this, mostri che si può lockare anche su un qualsiasi altro oggetto, tipo un attributo della classe e che questo può essere utile per "evitare di lockare troppo". Però pensandoci può giusto accennare che la cosa si può fare, tipo
```java
public class MyClass
    private Object lock;

    public void method() {
        synchronized(lock) {
             ...
        }
    }
}
```
e poi dire di guardare l'esempio da tema d'esame, perchè lì secondo me si capisce benissimo!